tests/unit/test_synapse_workload.py

```diff
@@ -61,3 +62,65 @@ def test_enable_metrics_error(monkeypatch: pytest.MonkeyPatch):
     monkeypatch.setattr(container_mock, "pull", pull_mock)
     with pytest.raises(synapse.WorkloadError, match=error_message):
         synapse.enable_metrics(container_mock)
+
+
+def test_enable_saml_success(harness_with_saml: Harness, monkeypatch: pytest.MonkeyPatch):
+    """
+    arrange: set mock container with file.
+    act: change the configuration file.
+    assert: new configuration file is pushed and SAML is enabled.
+    """
+    harness = harness_with_saml
+    config_content = """
+    listeners:
+        - type: http
+          port: 8080
+          bind_addresses:
+            - "::"
+    """
+    text_io_mock = io.StringIO(config_content)
+    pull_mock = Mock(return_value=text_io_mock)
+    push_mock = MagicMock()
+    container_mock = MagicMock()
```

Comment:
Any thoughts @weiiwang01 about using the new way to access the container without needing to mock?

Reply:
Yes, it's possible to use the new feature in the ops testing harness to prepare files in the container filesystem.
```diff
diff --git a/tests/unit/test_synapse_workload.py b/tests/unit/test_synapse_workload.py
index 816686e..c035774 100644
--- a/tests/unit/test_synapse_workload.py
+++ b/tests/unit/test_synapse_workload.py
@@ -82,17 +82,11 @@ def test_enable_saml_success(harness_with_saml: Harness, monkeypatch: pytest.Mon
             - "::"
           x_forwarded: false
     """
-    text_io_mock = io.StringIO(config_content)
-    pull_mock = Mock(return_value=text_io_mock)
-    push_mock = MagicMock()
-    container_mock = MagicMock()
-    monkeypatch.setattr(container_mock, "pull", pull_mock)
-    monkeypatch.setattr(container_mock, "push", push_mock)
-
-    synapse.enable_saml(container_mock, harness.charm._charm_state)
+    container_root = harness.get_filesystem_root("synapse")
+    synapse_config_file = container_root / "data/homeserver.yaml"
+    synapse_config_file.write_text(config_content, encoding="utf-8")
 
-    assert pull_mock.call_args[0][0] == SYNAPSE_CONFIG_PATH
-    assert push_mock.call_args[0][0] == SYNAPSE_CONFIG_PATH
+    synapse.enable_saml(harness.charm.unit.get_container("synapse"), harness.charm._charm_state)
     saml_relation = harness.model.get_relation("saml")
     assert saml_relation
     saml_relation_data = harness.get_relation_data(saml_relation.id, "saml-integrator")
@@ -115,7 +109,9 @@ def test_enable_saml_success(harness_with_saml: Harness, monkeypatch: pytest.Mon
             },
         },
     }
-    assert push_mock.call_args[0][1] == yaml.safe_dump(expected_config_content)
+    assert synapse_config_file.read_text(encoding="utf-8") == yaml.safe_dump(
+        expected_config_content
+    )
 
 
 def test_enable_saml_error(harness_with_saml: Harness, monkeypatch: pytest.MonkeyPatch):
```

Reply:
Thanks, the test was changed now.